TestF77Callback::test_all[t] fails: fatal error: threads.h: No such file or directory

<!-- Please describe the issue in detail here, and fill in the fields below -->

### Reproducing code example:

<!-- A short code example that reproduces the problem/missing feature. It should be
self-contained, i.e., possible to run as-is via 'python myproblem.py' -->

```
pytest -sv numpy/f2py/tests/test_callback.py -k 'test_all[t]' -r s -x
```

### Error message:

<!-- If you are reporting a segfault please include a GDB traceback, which you
can generate by following
https://github.com/numpy/numpy/blob/master/doc/source/dev/development_environment.rst#debugging -->

<!-- Full error message, if any (starting from line Traceback: ...) -->

```
E                                  /tmp/tmphz3v9phb/src.linux-x86_64-3.9/_test_ext_module_5403module.c:75:10: fatal error: threads.h: No such file or directory
E                                     75 | #include <threads.h>
E                                        |          ^~~~~~~~~~~
E                                  compilation terminated.
```

### NumPy/Python version information:

<!-- Output from 'import sys, numpy; print(numpy.__version__, sys.version)' -->

```
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.21.0.dev0+438.gbf4ac1f81 3.9.1 | packaged by conda-forge | (default, Jan 10 2021, 02:55:42) 
[GCC 9.3.0]
```

### Additional information

```
$ ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1.2) 2.27
Copyright (C) 2018 Free Software Foundation, Inc.
```

C11 threads.h appears to be supported starting from glibc 2.28, see [here](https://sourceware.org/bugzilla/show_bug.cgi?id=14092#c10).

Undefining CPP macro F2PY_THREAD_LOCAL_DECL makes the test pass.

A solution is not to define F2PY_THREAD_LOCAL_DECL for glibc version < 2.28 in numpy/f2py/cfuncs.py .

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -550,8 +550,11 @@ cppmacros["F2PY_THREAD_LOCAL_DECL"] = """\
 #elif defined(__STDC_VERSION__) \\
       && (__STDC_VERSION__ >= 201112L) \\
       && !defined(__STDC_NO_THREADS__) \\
-      && (!defined(__GLIBC__) || __GLIBC_MINOR__>12)
-/* see https://github.com/numpy/numpy/pull/18180 discussion for details */
+      && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12))
+/* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
+   see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
+   so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence
+   of `threads.h` when using an older release of glibc 2.12 */
 #include <threads.h>
 #define F2PY_THREAD_LOCAL_DECL thread_local
 #elif defined(__GNUC__) \\

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -548,8 +548,10 @@ cppmacros["F2PY_THREAD_LOCAL_DECL"] = """\
 #if defined(_MSC_VER)
 #define F2PY_THREAD_LOCAL_DECL __declspec(thread)
 #elif defined(__STDC_VERSION__) \\
-      && (__STDC_VERSION__ == 201112L) \\
-      && !defined(__STDC_NO_THREADS__)
+      && (__STDC_VERSION__ >= 201112L) \\
+      && !defined(__STDC_NO_THREADS__) \\
+      && (!defined(__GLIBC__) || __GLIBC_MINOR__>12)
+/* see https://github.com/numpy/numpy/pull/18180 discussion for details */
 #include <threads.h>
 #define F2PY_THREAD_LOCAL_DECL thread_local
 #elif defined(__GNUC__) \\

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -550,7 +550,7 @@ cppmacros["F2PY_THREAD_LOCAL_DECL"] = """\
 #elif defined(__STDC_VERSION__) \\
       && (__STDC_VERSION__ >= 201112L) \\
       && !defined(__STDC_NO_THREADS__) \\
-      && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12))
+      && (!defined(__GLIBC__) || __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ > 12))
 /* __STDC_NO_THREADS__ was first defined in a maintenance release of glibc 2.12,
    see https://lists.gnu.org/archive/html/commit-hurd/2012-07/msg00180.html,
    so `!defined(__STDC_NO_THREADS__)` may give false positive for the existence

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -549,7 +549,8 @@ cppmacros["F2PY_THREAD_LOCAL_DECL"] = """\
 #define F2PY_THREAD_LOCAL_DECL __declspec(thread)
 #elif defined(__STDC_VERSION__) \\
       && (__STDC_VERSION__ >= 201112L) \\
-      && !defined(__STDC_NO_THREADS__)
+      && !defined(__STDC_NO_THREADS__) \\
+      && (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 28)
 #include <threads.h>
 #define F2PY_THREAD_LOCAL_DECL thread_local
 #elif defined(__GNUC__) \\

--- a/numpy/f2py/cfuncs.py
+++ b/numpy/f2py/cfuncs.py
@@ -548,9 +548,8 @@ cppmacros["F2PY_THREAD_LOCAL_DECL"] = """\
 #if defined(_MSC_VER)
 #define F2PY_THREAD_LOCAL_DECL __declspec(thread)
 #elif defined(__STDC_VERSION__) \\
-      && (__STDC_VERSION__ >= 201112L) \\
-      && !defined(__STDC_NO_THREADS__) \\
-      && (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 28)
+      && (__STDC_VERSION__ == 201112L) \\
+      && !defined(__STDC_NO_THREADS__)
 #include <threads.h>
 #define F2PY_THREAD_LOCAL_DECL thread_local
 #elif defined(__GNUC__) \\

